closure_captures.md

```diff
@@ -0,0 +1,319 @@
+provides a visitor/walker that can traverse the tree in execution order.
+We can use it and override `visit_*` methods, to have a walker that meets our needs.
+
+
+#### [`upvars_mentioned`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/arena/struct.Arena.html#structfield.upvars_mentioned)
+Generated in [`librustc_passes`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_passes/upvars.rs.html#82).
+
+Here we implement the instravisit visitor that can generate a list of local root variables by looking at `PatKind::Binding` (**TODO: explain**).
+
+Once that is done, we walk the body again to look for
+- [`Path`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Path.html) that represents a variable use (`Res::Local`).
+    Path isn't something like `a.b` but just represents some variable use.
+    If the variable isn't a local variable we add it to the list of upvars and map it to the Path span.
+- `Expr` that represents a Closure, i.e. handle the case of nested closures. We can query `upvars_mentioned` for the nested closure and include all the upvars returned, as part of captures for the current (enclosing) closure.
+
+
+#### [`TypeckTables::closure_captures`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TypeckTables.html#structfield.closure_captures)
+
+Generated in [`librustc_typeck`](https://github.com/rust-lang/rust/blob/9bdd2db3a60176012f4dc240eea02d615cc60061/src/librustc_typeck/check/upvar.rs#L115).
+
+This map for a given closure is very similar to upvars_mentioned. Instead of storing span we store the pair of `(closure_def_id, root_var_hir_id)`.
+
+We iterate over `upvars_mentioned` for a given `closure_def_id` to get captured root variable HirId and generate an UpvarId for them and build the map.
+
+#### [`TypeckTables::upvar_capture_map`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TypeckTables.html#structfield.upvar_capture_map)
+
+Generated in [`librustc_typeck`](https://github.com/rust-lang/rust/blob/9bdd2db3a60176012f4dc240eea02d615cc60061/src/librustc_typeck/check/upvar.rs#L115).
+
+We implement another `intravist::Vistor` to analyse closures within a scope.
+When we see a closure within the body of the scope, we walk the body of the closure, when we are done walking the closure's body we start the analysis on the closure.
+(We need to walk first, to handle nested closures.)
+
+
+To begin the analysis we use hints like `mut` for c5, `move` in c6, to guess the initial borrow kind of each capture.
+Once the guesses are set, we create an ExprUseVistior that visits each expression within the closure body.
+ExprUseVisitor calls into MemCategorization which returns the access information about the expression as a `Place`.
+
+ExprUseVisitor delegates the work of marking a Place as consume/borrow to a callback. In our case the `Delegate` trait is implemented
```

Comment:
I would say it like this:

ExprUseVisitor invokes methods on the delegate when it finds a Place being consumed or borrowed. The delegate has the job of deciding what to do with that information. In this case, the delegate is the upvar analysis, which "upgrades" the kind of borrow needed depending on the accesses it sees (all reads == shared borrow, writes == mutable borrow, moves == take ownership).

(Currently, ExprUseVisitor is I think only used in one place, actually, so we could conceivably "inline it" into the upvar analysis, but it's kind of a nice separation..?)

Reply:
I'll update this tonight. We do use ExprUseVisitor in clippy, but I think separation makes it easier to understand from code prespective as well.